security/pkg/pki/ra/k8s_ra.go

```diff
@@ -0,0 +1,160 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ra
+
+import (
+	"fmt"
+	"time"
+
+	cert "k8s.io/api/certificates/v1beta1"
+	certclient "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+
+	"istio.io/istio/security/pkg/k8s/chiron"
+	caerror "istio.io/istio/security/pkg/pki/error"
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+// IstioRA : Main type definition of IstioRA
+type IstioRA struct {
+	keyCertBundle util.KeyCertBundle
+	raOpts        *IstioRAOptions
+	csrInterface  certclient.CertificatesV1beta1Interface
+	// TODO: Create a watcher for the rootCA file
+}
+
+// IstioRAOptions : Configuration Options for the IstioRA
+type IstioRAOptions struct {
+	DefaultCertTTL time.Duration
+	MaxCertTTL     time.Duration
+	caCertFile     string
+	caSigner       string
+}
+
+// NewK8sRAOptions : Initialize Kubernates Options
+func NewK8sRAOptions(defaultCertTTL, maxCertTTL time.Duration,
+	caCertFile, caSigner string) *IstioRAOptions {
+	raOptions := &IstioRAOptions{
+		DefaultCertTTL: defaultCertTTL,
+		MaxCertTTL:     maxCertTTL,
+		caCertFile:     caCertFile,
+		caSigner:       caSigner}
+	return raOptions
+}
+
+// NewK8sRA : Create a RA that interfaces with K8S CSR CA
+func NewK8sRA(raOpts *IstioRAOptions, csrInterface certclient.CertificatesV1beta1Interface) (*IstioRA, error) {
+	keyCertBundle, err := util.NewKeyCertBundleWithRootCertFromFile(raOpts.caCertFile)
+	if err != nil {
+		return nil, caerror.NewError(caerror.CSRError, fmt.Errorf("error creating Certificate Bundle for k8s CA"))
+	}
+	istioRA := &IstioRA{csrInterface: csrInterface,
+		raOpts:        raOpts,
+		keyCertBundle: keyCertBundle}
+	return istioRA, nil
+}
+
+// Validate all SAN extensions in csrPEM match authenticated identities
+func (ra *IstioRA) validateCSR(csrPEM []byte, subjectIDs []string) bool {
```

Comment:
I don't get it. How can we have multiple authenticated identities ?

Reply:
I am unsure about this as well. 
When Istiod functions as a CA, it tries to authenticate (and authorize) the CSR request using one of many authenticators functions. Each of this yield multiple identities. Adding @howardjohn for comment since he refactored this code most recently.

This PR ensures that the Istio-agent embeds its identity as a SAN in the CSR Request. This piece of code just ensures that the SAN retrieved from the CSR Request matches one of the identities extracted.